src/v/kafka/server/group.cc

```diff
@@ -1674,13 +1674,60 @@ void group::insert_prepared(prepared_tx tx) {
 
 ss::future<cluster::commit_group_tx_reply>
 group::commit_tx(cluster::commit_group_tx_request r) {
-    // doesn't make sense to fence off a commit because transaction
-    // manager has already decided to commit and acked to a client
-
     if (_partition->term() != _term) {
         co_return make_commit_tx_reply(cluster::tx_errc::stale);
     }
 
+    auto fence_it = _fence_pid_epoch.find(r.pid.get_id());
+    if (fence_it == _fence_pid_epoch.end()) {
+        vlog(
+          _ctx_txlog.warn,
+          "Can't prepare tx: fence with pid {} isn't set",
+          r.pid);
+        co_return make_commit_tx_reply(cluster::tx_errc::request_rejected);
+    }
+    if (r.pid.get_epoch() != fence_it->second) {
+        vlog(
+          _ctx_txlog.trace,
+          "Can't prepare tx with pid {} - the fence doesn't match {}",
+          r.pid,
+          fence_it->second);
+        co_return make_commit_tx_reply(cluster::tx_errc::request_rejected);
+    }
+
+    auto txseq_it = _tx_seqs.find(r.pid.get_id());
+    if (txseq_it == _tx_seqs.end()) {
+        if (is_transaction_ga()) {
+            vlog(
+              _ctx_txlog.trace,
+              "can't find a tx {}, probably already comitted",
+              r.pid);
+            co_return make_commit_tx_reply(cluster::tx_errc::none);
+        }
```

Comment:
so here if txn-ga is not enabled, then we _fallthrough_ and continue with prepare?

Reply:
correct, this is expected